inc/zoo/swar/math.h

```diff
@@ -0,0 +1,69 @@
+#pragma once
+#include "SWAR.h"
+
+namespace zoo::math {
+
+template <typename IntegerType = size_t>
+constexpr static
+std::enable_if_t<std::is_integral_v<IntegerType>, bool>
+isPowerOfTwo(IntegerType x) noexcept {
+    return x && (x & (x - 1)) == 0;
+}
+
+template <typename IntegerType = size_t, IntegerType X>
+constexpr static
+std::enable_if_t<std::is_integral_v<IntegerType>, bool>
+isPowerOfTwo() noexcept {
+    return isPowerOfTwo(X);
+}
+static_assert(isPowerOfTwo<int, 4>());
+
+
+template <size_t N, typename IntegerType = size_t>
+constexpr static
+std::enable_if_t<
+    std::is_integral_v<IntegerType> &&
+    isPowerOfTwo<size_t, N>(), size_t>
+moduloPowerOfTwo(IntegerType x) noexcept {
+    return x & (N - 1);
+}
+
+static_assert(moduloPowerOfTwo<4>(0) == 0);
+static_assert(moduloPowerOfTwo<8>(9) == 1);
+static_assert(moduloPowerOfTwo<4096>(4097) == 1);
+
+}
+
+namespace zoo::swar {
+template <typename S>
+constexpr static auto subtractOneUnsafe(S x) noexcept {
+    constexpr auto Ones = S::LeastSignificantBit;
+    auto x_minus_1 = S{x.value() - Ones};
+    return x_minus_1;
+}
+// todo subtract K unsafe using BitmaskMaker
+// todo subtract K "saturated" using BitmaskMaker
+
+template <typename S>
+constexpr static auto isPowerOfTwo(S x) noexcept {
+    constexpr auto NBits = S::NBits;
+    using T = typename S::type;
+    auto greater_than_0 = greaterEqual_MSB_off(x, S{0});
+    auto x_minus_1 = subtractOneUnsafe(x);
+    auto zero = equals(S{x_minus_1.value() & x.value()}, S{0});
+    return greater_than_0 & zero;
+}
+
+template <size_t N, typename S>
+constexpr static
+std::enable_if_t<zoo::math::isPowerOfTwo<size_t, N>(), S>
+moduloPowerOfTwo(const S x) noexcept {
+    constexpr auto N_minus_1 = N - 1;
+    constexpr auto N_in_lanes = zoo::meta::BitmaskMaker<typename S::type, N_minus_1, S::NBits>::value;
+    auto y = x.value() & N_in_lanes;
+    return S{y};
+}
+
```

Comment:
The way this is formulated is perhaps not useful:
For integral types, when the compiler can prove the modulus is a power of two, the compiler will, for sure, use bitwise conjunction rather than the assembler for the modulo instruction.
There remains the question of when the compiler doesn't know whether the modulus is a power of two, this would be only at runtime, and then, programmers have explicitly accounted for this improvement since ages, I don't think this way of going about it will give them an incentive to use this primitive.
This is useful for "weird" types that represent integers, like multi-precision integers, I don't know, that do not satisfy the "Integral" property required by `isPowerOfTwo`.
There is one aspect that **I really like** aobut this code: the structural guarantee that you won't pass a non-power-of-two as template argument, I see myself making some code with lane sizes not powers of two, and in some remote part of the code there is an assumption broken of the lane size being a powr of two, invoking bitwise conjuction, for "hilarity".